src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.h

```diff
@@ -46,10 +47,43 @@
 struct grpc_tls_certificate_provider
     : public grpc_core::RefCounted<grpc_tls_certificate_provider> {
  public:
+  // The pointer value \a type is used to uniquely identify a creds
+  // implementation for down-casting purposes. Every creds implementation should
+  // use a unique string instance for every creds instance of that creds
+  // implementation.
+  explicit grpc_tls_certificate_provider(const char* type) : type_(type) {}
+
   virtual grpc_pollset_set* interested_parties() const { return nullptr; }
 
   virtual grpc_core::RefCountedPtr<grpc_tls_certificate_distributor>
   distributor() const = 0;
+
+  // Compares this grpc_tls_certificate_provider object with \a other.
+  // If this method returns 0, it means that gRPC can treat the two channel
+  // credentials as effectively the same. This method is used to compare
+  // `grpc_tls_certificate_provider` objects when they are present in
+  // channel_args. One important usage of this is when channel args are used in
+  // SubchannelKey, which leads to a useful property that allows subchannels to
+  // be reused when two different `grpc_tls_certificate_provider` objects are
+  // used but they compare as equal (assuming other channel args match).
+  int cmp(const grpc_tls_certificate_provider* other) const {
+    GPR_ASSERT(other != nullptr);
+    // Intentionally uses grpc_core::QsortCompare instead of strcmp as a safety
+    // against different grpc_tls_certificate_provider types using the same
+    // name.
```

Comment:
Sorry, I don't get the benefit of using `grpc_core::QsortCompare` over `strcmp`.  If the names are the same, I think they will both return 0?

Reply:
So, the worry here is that what if by mistake, we use the same name for two different implementations? In that case, we would hope that the implementations atleast use different definitions of that string. We are just being defensive here.